tests/TestCollectors.py

```diff
@@ -191,9 +200,19 @@ def test_collector_metrics(self):
                 t.join(timeout=5)
 
             # we don't want to have any port locks if prometheus server thread is not shutting down
-            random_prometheus_port += 1
+            self.random_prometheus_port += 1
             REGISTRY.unregister(collector_instance)
 
+        for collector in metrics_yaml.keys():
+            print("\nTesting " + collector)
+            if collector == 'VMStatsCollector':
+                for rubric in metrics_yaml[collector]['metrics']:
+                    os.environ['RUBRIC'] = rubric
+                    print("Testing rubric:", rubric)
+                    collector_testrun(rubric)
+            else:
+                collector_testrun()
```

Comment:
Please split the testing in two dedicated testable units
- with rubric
- without rubric

You have to define a really big inner scope function to accomplish doing everything in here and it will continue to be hard to read.

Reply:
I agree. But I wouldn't copy paste code just for testing this tiny difference. Let me remove the specific collector in there to make it a more general approach.